File: test-2025.9.2/编程/work2.cpp
#include <stdio.h>

#define N9 9
#define N4 4

int n;  
int board[9][9];  


int isValid(int row, int col, int num) {
    int blockSize = (n == 9) ? 3 : 2;
    for (int i = 0; i < n; i++) {
        if (board[row][i] == num) return 0;
        if (board[i][col] == num) return 0;
    }
    int startRow = (row / blockSize) * blockSize;
    int startCol = (col / blockSize) * blockSize;
    for (int i = 0; i < blockSize; i++) {
        for (int j = 0; j < blockSize; j++) {
            if (board[startRow + i][startCol + j] == num) return 0;
        }
    }
    return 1;
}


int solveSudoku() {
    for (int row = 0; row < n; row++) {
        for (int col = 0; col < n; col++) {
            if (board[row][col] == 0) {
                for (int num = 1; num <= n; num++) {
                    if (isValid(row, col, num)) {
                        board[row][col] = num;
                        if (solveSudoku()) return 1;
                        board[row][col] = 0; 
                    }
                }
                return 0; 
            }
        }
    }
    return 1; 
}

int main() {
    scanf("%d", &n);
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            scanf("%d", &board[i][j]);
        }
    }

    if (solveSudoku()) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                printf("%d", board[i][j]);
                if (j != n - 1) printf(" ");
            }
            printf("\n");
        }
    } else {
        printf("No solution\n");
    }
    return 0;
}

File: work_2/content/编程测试题/work2.py
def is_valid(board, row, col, num, size):
    for c in range(size):
        if board[row][c] == num:
            return False
    for r in range(size):
        if board[r][col] == num:
            return False
    sub_size = int(size ** 0.5)
    start_row, start_col = row - row % sub_size, col - col % sub_size
    for r in range(start_row, start_row + sub_size):
        for c in range(start_col, start_col + sub_size):
            if board[r][c] == num:
                return Falserk
    return True

def solve_sudoku(board, size):
    for row in range(size):
        for col in range(size):
            if board[row][col] == 0:
                for num in range(1, size + 1):
                    if is_valid(board, row, col, num, size):
                        board[row][col] = num
                        if solve_sudoku(board, size):
                            return True
                        board[row][col] = 0
                return False
    return True

if __name__ == "__main__":
    size = int(input().strip())
    board = [list(map(int, input().strip().split())) for _ in range(size)]
    solve_sudoku(board, size)
    for row in board:
        print(" ".join(map(str, row)))
